### PR TITLE
feat: Tessera models

### PR DESCRIPTION
**Pull request by isaaccorley**
## Summary

Adds the Tessera pixel time series foundation model to torchgeo, ported from [pixelverse](https://github.com/developmentseed/pixelverse/blob/main/src/pixelverse/models/tessera.py).

## Changes

- **New model**: `torchgeo/models/tessera.py`
  - `Tessera`: Combined S1+S2 encoder with 128-dim output embedding
  - `TransformerEncoder`: Transformer encoder for pixel time series
  - `TemporalAwarePooling`: Attention-based temporal pooling
  - `TemporalPositionalEncoder`: Sinusoidal day-of-year positional encoding
  - `_TesseraTransforms`: Normalization transforms for inference
  - Factory functions: `tessera()`, `tessera_s2_encoder()`, `tessera_s1_encoder()`
  - Weight enums: `Tessera_Weights`, `Tessera_S2_Encoder_Weights`, `Tessera_S1_Encoder_Weights`

- **Tests**: `tests/models/test_tessera.py` (12 tests, 100% coverage)

- **Documentation**: 
  - `docs/api/models/tessera.rst`
  - Added to `docs/api/models.rst` toctree
  - Added weight entries to `docs/api/weights/agnostic.csv`

## Model Details

Tessera expects input shape `(B, seq_len, 14)` where:
- Channels 0-9: Sentinel-2 bands (B2-B12)
- Channel 10: Sentinel-2 day of year
- Channels 11-12: Sentinel-1 VV and VH
- Channel 13: Sentinel-1 day of year

Pretrained weights hosted at [isaaccorley/tessera](https://huggingface.co/isaaccorley/tessera) on HuggingFace.

## References

- Paper: https://arxiv.org/abs/2505.05416
- Original implementation: https://github.com/developmentseed/pixelverse

@FrankFeng-23 @sk818 @avsm @rbnyng

### CODE REVIEW

**Comment by adamjstewart** [comment]
https://github.com/developmentseed/pixelverse is unlicensed, why not instead port from https://github.com/ucam-eo/tessera? We can literally copy-n-paste the model and retain the copyright since it's already under an MIT license.

**Comment by isaaccorley** [comment]
> https://github.com/developmentseed/pixelverse is unlicensed, why not instead port from https://github.com/ucam-eo/tessera? We can literally copy-n-paste the model and retain the copyright since it's already under an MIT license.

Thanks for catching that, I'll add an MIT license to pixelverse.

I made some small optimizations to TESSERA code since it was mixing numpy, math, and torch code for whatever reason but sure we can copy paste directly from source for clarity.

File: docs/api/weights/agnostic.csv [hunk 1]
```diff
@@ -7,3 +7,4 @@ DOFALarge16_Weights.DOFA_MAE,`link <https://github.com/zhu-xlab/DOFA>`__,`link <
 Panopticon_Weights.VIT_BASE14,`link <https://github.com/Panopticon-FM/panopticon>`__,`link <https://www.arxiv.org/abs/2503.10845>`__,Apache-2.0,implicit,,explicit,-,56.3,96.7,96.4,61.7,96.4,95.2,92.6,79.6,59.3,52.6,78.1
 ResNet50_Weights.FMOW_RGB_GASSL,`link <https://github.com/sustainlab-group/geography-aware-ssl>`__,`link <https://arxiv.org/abs/2011.09980>`__,-,implicit,-,-,,,,,,,,,,,,
 ScaleMAE_ViTLarge16_Weights.FMOW_RGB_SCALEMAE,`link <https://github.com/bair-climate-initiative/scale-mae>`__,`link <https://arxiv.org/abs/2212.14532>`__,CC-BY-NC-4.0,explicit,-,-,,,,,,,,,,,
+Tessera_Weights.TESSERA,`link <https://github.com/ucam-eo/tessera>`__,`link <https://arxiv.org/abs/2506.20380>`__,MIT,-,explicit,implicit,explicit,,,,,,,,,,,
```
Comment (adamjstewart) [inline docs/api/weights/agnostic.csv]: * Spatial: I don't see anything related to spatial resolution invariance in the model architecture (explicit) or in the training data (implicit). The model is solely trained on 10 m resolution imagery and only supports a single pixel anyway.
* Temporal: I see learnable DoY position encoders (explicit)
* Spectral: I don't see anything related to spectral channel invariance. There are two separate models (Sentinel-1 and Sentinel-2) combined via MLP. It is very unlikely that the model could support any other sensors.

I also think you have an extra column.

```suggestion
Tessera_Weights.TESSERA,`link <https://github.com/ucam-eo/tessera>`__,`link <https://arxiv.org/abs/2506.20380>`__,MIT,-,explicit,-,,,,,,,,,,,
```

File: docs/api/weights/sentinel1.csv [hunk 2]
```diff
@@ -17,4 +17,5 @@ ViTHuge14_Weights.SENTINEL1_GRD_FGMAE,2,`link <https://github.com/zhu-xlab/FGMAE
 ViTSmall14_DINOv2_Weights.SENTINEL1_GRD_SOFTCON, 2,`link <https://github.com/zhu-xlab/softcon>`__,`link <https://arxiv.org/abs/2405.20462>`__,"CC-BY-4.0"
 ViTBase14_DINOv2_Weights.SENTINEL1_GRD_SOFTCON, 2,`link <https://github.com/zhu-xlab/softcon>`__,`link <https://arxiv.org/abs/2405.20462>`__,"CC-BY-4.0"
 Swin_V2_B_Weights.SENTINEL1_MI_SATLAS,2,`link <https://github.com/allenai/satlas>`__,`link <https://arxiv.org/abs/2211.15660>`__,ODC-BY
-Swin_V2_B_Weights.SENTINEL1_SI_SATLAS,2,`link <https://github.com/allenai/satlas>`__,`link <https://arxiv.org/abs/2211.15660>`__,ODC-BY
\ No newline at end of file
+Swin_V2_B_Weights.SENTINEL1_SI_SATLAS,2,`link <https://github.com/allenai/satlas>`__,`link <https://arxiv.org/abs/2211.15660>`__,ODC-BY
+Tessera_Weights.TESSERA_SENTINEL1_ENCODER,3,`link <https://github.com/ucam-eo/tessera>`__,`link <https://arxiv.org/abs/2506.20380>`__,"MIT"
```
Comment (adamjstewart) [inline docs/api/weights/sentinel1.csv]: Can you add an EOL char so we don't keep editing the last line with every PR?

File: torchgeo/models/tessera.py [hunk 3]
```diff
@@ -0,0 +1,392 @@
+# Copyright (c) TorchGeo Contributors. All rights reserved.
+# Licensed under the MIT License.
+#
+# Based on the original code: https://github.com/ucam-eo/tessera
```
Comment (adamjstewart) [inline torchgeo/models/tessera.py:5]: How heavily based? Do we need to preserve copyright information?

File: torchgeo/models/tessera.py [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+
+from typing import Any
+
+import torch
+import torchvision.transforms.v2 as T
+from einops import rearrange
+from torch import Tensor, nn
+from torchvision.models._api import Weights, WeightsEnum
+
+_S2_BAND_MEAN = [
```
Comment (adamjstewart) [inline torchgeo/models/tessera.py:18]: Maybe add a comment pointing to where these values come from in case they change?

File: torchgeo/models/tessera.py [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+_S1_BAND_STD = [1871.2334, 1726.0670]
+_TESSERA_S2_MEAN = [*_S2_BAND_MEAN, 0.0]
+_TESSERA_S2_STD = [*_S2_BAND_STD, 1.0]
+_TESSERA_S1_MEAN = [*_S1_BAND_MEAN, 0.0]
+_TESSERA_S1_STD = [*_S1_BAND_STD, 1.0]
+_TESSERA_MEAN = _TESSERA_S2_MEAN + _TESSERA_S1_MEAN
+_TESSERA_STD = _TESSERA_S2_STD + _TESSERA_S1_STD
+
+
+class _PixelTimeSeriesNormalize(T.Normalize):
```
Comment (adamjstewart) [inline torchgeo/models/tessera.py]: If we decide to replace Kornia with torchvision everywhere, this could be a useful transform.

File: torchgeo/models/tessera.py [hunk 6]
```diff
@@ -0,0 +1,392 @@
...
+
+
+class Tessera_Weights(WeightsEnum):  # type: ignore[misc]
+    """Tessera model weights.
+
+    .. versionadded:: 0.9
+    """
+
+    TESSERA = Weights(
+        url='https://hf.co/isaaccorley/tessera/resolve/51afe75b724d387ef9fcb6f6e090a5be0b906919/model.pt',
```
Comment (adamjstewart) [inline torchgeo/models/tessera.py]: I try to include the first 8 digits of the sha256sum in the filename so that `torch.hub.load_from_state_dict_url` (or whatever its called) can automatically do checksum verification. Not required, but if you're planning on converting all models to pt2 someday, something to keep in mind.

File: torchgeo/models/tessera.py [hunk 7]
```diff
@@ -0,0 +1,392 @@
...
+        transforms=_tessera_transforms,
+        meta={
+            'dataset': 'TESSERA',
```
Comment (adamjstewart) [inline torchgeo/models/tessera.py:300]: I actually don't see any discussion or name of the dataset TESSERA was trained on.

**Comment by avsm** [comment]
Awesome! We're happy to put the weights onto a HF org later on as well, to help with versioning, but for now this is fantastic. Thanks @isaaccorley!

File: torchgeo/models/tessera.py (hunk 7 above)
Comment (isaaccorley) [reply torchgeo/models/tessera.py:300]: I don't either, I don't think they actually gave it a name judging from the paper. Would you recommend just removing this field?

Comment (adamjstewart) [reply torchgeo/models/tessera.py:300]: Nah, it's fine

File: torchgeo/models/tessera.py (hunk 3 above)
Comment (adamjstewart) [reply torchgeo/models/tessera.py:5]: If so we need to add:
```
Copyright (c) 2025 Frank Feng
```

File: docs/api/weights/agnostic.csv [hunk 8]
```diff
@@ -7,3 +7,4 @@ DOFALarge16_Weights.DOFA_MAE,`link <https://github.com/zhu-xlab/DOFA>`__,`link <
 Panopticon_Weights.VIT_BASE14,`link <https://github.com/Panopticon-FM/panopticon>`__,`link <https://www.arxiv.org/abs/2503.10845>`__,Apache-2.0,implicit,,explicit,-,56.3,96.7,96.4,61.7,96.4,95.2,92.6,79.6,59.3,52.6,78.1
 ResNet50_Weights.FMOW_RGB_GASSL,`link <https://github.com/sustainlab-group/geography-aware-ssl>`__,`link <https://arxiv.org/abs/2011.09980>`__,-,implicit,-,-,,,,,,,,,,,,
 ScaleMAE_ViTLarge16_Weights.FMOW_RGB_SCALEMAE,`link <https://github.com/bair-climate-initiative/scale-mae>`__,`link <https://arxiv.org/abs/2212.14532>`__,CC-BY-NC-4.0,explicit,-,-,,,,,,,,,,,
+Tessera_Weights.TESSERA,`link <https://github.com/ucam-eo/tessera>`__,`link <https://arxiv.org/abs/2506.20380>`__,MIT,-,explicit,-,,,,,,,,,,,
```
Comment (adamjstewart) [inline docs/api/weights/agnostic.csv]: @avsm can you confirm if the model weights are MIT like the code or CC-BY-4.0 like the data?

Comment (adamjstewart) [reply docs/api/weights/agnostic.csv]: Ping @avsm

File: docs/api/weights/agnostic.csv [hunk 9]
```diff
@@ -7,3 +7,4 @@ DOFALarge16_Weights.DOFA_MAE,`link <https://github.com/zhu-xlab/DOFA>`__,`link <
 Panopticon_Weights.VIT_BASE14,`link <https://github.com/Panopticon-FM/panopticon>`__,`link <https://www.arxiv.org/abs/2503.10845>`__,Apache-2.0,implicit,,explicit,-,56.3,96.7,96.4,61.7,96.4,95.2,92.6,79.6,59.3,52.6,78.1
 ResNet50_Weights.FMOW_RGB_GASSL,`link <https://github.com/sustainlab-group/geography-aware-ssl>`__,`link <https://arxiv.org/abs/2011.09980>`__,-,implicit,-,-,,,,,,,,,,,,
 ScaleMAE_ViTLarge16_Weights.FMOW_RGB_SCALEMAE,`link <https://github.com/bair-climate-initiative/scale-mae>`__,`link <https://arxiv.org/abs/2212.14532>`__,CC-BY-NC-4.0,explicit,-,-,,,,,,,,,,,
+Tessera_Weights.TESSERA,`link <https://github.com/ucam-eo/tessera>`__,`link <https://arxiv.org/abs/2506.20380>`__,CC-BY-4.0,-,explicit,-,,,,,,,,,,,
```
Comment (adamjstewart) [inline docs/api/weights/agnostic.csv]: Let's use "closed source" until the authors decide which license to release the data under.

**Comment by adamjstewart** [comment]
@FrankFeng-23 @sk818 @avsm @rbnyng here is our review paper: https://arxiv.org/abs/2601.13134

Let us know if there are any inaccuracies or we missed anything!

**Comment by FrankFeng-23** [comment]
The paper looks very good, but there's one point I think might need modification: We mentioned in the tessera [repo](https://github.com/ucam-eo/tessera):

-  "Next, download the model weights from Google Drive (please request for access) and place the .pt file in the tessera_infer/checkpoints directory:" 

 This means our checkpoint is also public. @avsm @sk818 might be able to clarify what the license is for our checkpoint?

**Comment by FrankFeng-23** [comment]
We remove "(please request for access)" as everyone can access it without asking for permission.

**Comment by isaaccorley** [comment]
@FrankFeng-23 this is all great, however a license needs to be defined to remove ambiguity for industry users to start using your embedding data product and model in their applications.

**Comment by FrankFeng-23** [comment]
yes! totally agree. we're discussing this issue internally and will work it out soon!

**Comment by adamjstewart** [comment]
Yes, "source available" would probably be more accurate than "closed source". Let us know when you decide on a specific license and we'll update both the paper and the TorchGeo docs!

**Comment by adamjstewart** [comment]
Also, there are a bunch of people discussing the paper on the [TorchGeo Slack](https://join.slack.com/t/torchgeo/shared_invite/zt-22rse667m-eqtCeNW0yI000Tl4B~2PIw) if any of you want to join to answer any specific questions about Tessera!

**Comment by sk818** [comment]
Hi Adam,
 We’ve updated the git repo to reflect CC0 license!


with kind regards

keshav




From: Adam J. Stewart ***@***.***>
Date: Wednesday, 21 January 2026 at 5:40 pm
To: torchgeo/torchgeo ***@***.***>
Cc: S. Keshav ***@***.***>, Mention ***@***.***>
Subject: Re: [torchgeo/torchgeo] feat: Tessera models (PR #3243)

[https://avatars.githubusercontent.com/u/12021217?s=20&v=4]adamjstewart left a comment (torchgeo/torchgeo#3243)<https://github.com/torchgeo/torchgeo/pull/3243#issuecomment-3780009196>

Yes, "source available" would probably be more accurate than "closed source". Let us know when you decide on a specific license and we'll update both the paper and the TorchGeo docs!

**Comment by adamjstewart** [comment]
For licenses I'm now seeing:

* model code: MIT
* model weights: ?
* train/inference data: ?
* embeddings: CC0-1.0

Is the train/inference data released anywhere or are there scripts to redownload it for reproducibility?

Also worth updating https://github.com/ucam-eo/tessera/issues/5 and https://github.com/ucam-eo/geotessera/issues/150, the license situation of geotessera is even more unclear (MIT vs. ISC).

**Comment by sk818** [comment]
I've updated the GitHub to make model weights CC0. Training data is from S1/S2 from ESA, so not ours to license and yes, we do provide a s1_s2_downloader script to access data from either Microsoft Planetary computer or AWS. I am not sure what 'inference data' refers to, since our inferences are embeddings, which are already CC0. We are aware of the ISC license issue for GeoTessera : this will be updated very shortly (today or tomorrow) to MIT.

**Comment by avsm** [comment]
And geotessera also updated to be MIT; sorry about the confusion! (I normally use ISC for my OCaml code, but MIT is more conventional so I switched to that to be consistent with TESSERA itself). https://github.com/ucam-eo/geotessera/commit/708191d6ee39366d3d45070f4f7f2975447639c2
